File: 2022-2023/sourcecode/lab/lab13/DaggerHiltROOMBasicsJava/app/src/main/java/pl/edu/uwr/pum/dagger_hilt_room_basicsjava/data/db/AppDatabase.java
package pl.edu.uwr.pum.dagger_hilt_room_basicsjava.data.db;


import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import pl.edu.uwr.pum.dagger_hilt_room_basicsjava.data.Student;

@Database(entities = {Student.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase{

        public abstract AppDao appDao();

        private static volatile AppDatabase INSTANCE;

        private static final int NUMBER_OF_THREADS = 4;
        public static final ExecutorService databaseWriteExecutor =
                Executors.newFixedThreadPool(NUMBER_OF_THREADS);

        public static AppDatabase getDatabase(final Context context) {
            if (INSTANCE == null) {
                synchronized (AppDatabase.class) {
                    if (INSTANCE == null) {
                        INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                                        AppDatabase.class, "student_database_java")
                                .build();
                    }
                }
            }
            return INSTANCE;
        }
}

File: 2022-2023/sourcecode/lab/lab13/DaggerHiltROOMBasicsKotlin/app/src/main/java/pl/edu/uwr/pum/dagger_hilt_room_basicskotlin/data/db/AppDatabase.kt
package pl.edu.uwr.pum.dagger_hilt_room_basicskotlin.data.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import pl.edu.uwr.pum.dagger_hilt_room_basicskotlin.data.Student

@Database(entities = [Student::class], version = 1)
abstract class AppDatabase() : RoomDatabase() {

    abstract fun appDao(): AppDao

    companion object{
        @Volatile private var INSTANCE: AppDatabase? = null

        fun getDatabase(context: Context): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "student_database_kotlin"
                ).build().also { INSTANCE = it }
                instance
            }
        }
    }
}